substrate/client/rpc-servers/src/lib.rs

```diff
@@ -23,246 +23,255 @@
-			let cfg2 = cfg.clone();
-			let svc = tower::service_fn(move |req: http::Request<hyper::body::Incoming>| {
-				let PerConnection {
-					methods,
-					service_builder,
-					metrics,
-					tokio_handle,
-					stop_handle,
-					rate_limit_whitelisted_ips,
-				} = cfg2.clone();
-
-				let proxy_ip =
-					if rate_limit_trust_proxy_headers { get_proxy_ip(&req) } else { None };
-
-				let rate_limit_cfg = if rate_limit_whitelisted_ips
-					.iter()
-					.any(|ips| ips.contains(proxy_ip.unwrap_or(ip)))
-				{
-					log::debug!(target: "rpc", "ip={ip}, proxy_ip={:?} is trusted, disabling rate-limit", proxy_ip);
-					None
-				} else {
-					if !rate_limit_whitelisted_ips.is_empty() {
-						log::debug!(target: "rpc", "ip={ip}, proxy_ip={:?} is not trusted, rate-limit enabled", proxy_ip);
-					}
-					rate_limit
+					_ = cfg.stop_handle.clone().shutdown() => break,
 				};
 
-				let is_websocket = ws::is_upgrade_request(&req);
-				let transport_label = if is_websocket { "ws" } else { "http" };
-
-				let middleware_layer = match (metrics, rate_limit_cfg) {
-					(None, None) => None,
-					(Some(metrics), None) => Some(
-						MiddlewareLayer::new().with_metrics(Metrics::new(metrics, transport_label)),
-					),
-					(None, Some(rate_limit)) =>
-						Some(MiddlewareLayer::new().with_rate_limit_per_minute(rate_limit)),
-					(Some(metrics), Some(rate_limit)) => Some(
-						MiddlewareLayer::new()
-							.with_metrics(Metrics::new(metrics, transport_label))
-							.with_rate_limit_per_minute(rate_limit),
-					),
+				let RpcSettings {
+					batch_config,
+					max_connections,
+					max_payload_in_mb,
+					max_payload_out_mb,
+					max_buffer_capacity_per_connection,
+					max_subscriptions_per_connection,
+					rpc_methods,
+					rate_limit_trust_proxy_headers,
+					rate_limit_whitelisted_ips,
+					host_filter,
+					cors,
+					rate_limit,
+				} = rpc_cfg;
+
+				let http_middleware = tower::ServiceBuilder::new()
+					.option_layer(host_filter)
+					// Proxy `GET /health, /health/readiness` requests to the internal
+					// `system_health` method.
+					.layer(NodeHealthProxyLayer::default())
+					.layer(cors);
+
+				let mut builder = jsonrpsee::server::Server::builder()
+					.max_request_body_size(max_payload_in_mb.saturating_mul(MEGABYTE))
+					.max_response_body_size(max_payload_out_mb.saturating_mul(MEGABYTE))
+					.max_connections(max_connections)
+					.max_subscriptions_per_connection(max_subscriptions_per_connection)
+					.enable_ws_ping(
+						PingConfig::new()
+							.ping_interval(Duration::from_secs(30))
+							.inactive_limit(Duration::from_secs(60))
+							.max_failures(3),
+					)
+					.set_http_middleware(http_middleware)
+					.set_message_buffer_capacity(max_buffer_capacity_per_connection)
+					.set_batch_request_config(batch_config)
+					.custom_tokio_runtime(cfg.tokio_handle.clone())
+					.set_id_provider(RandomStringIdProvider::new(16));
+
+				if let Some(provider) = id_provider2.take() {
```

Comment:
Hey, if you take a look we are now using a `tokio::spawn task` for each RPC endpoint and that's why we need `take() and clone()` because it's used in a loop

Reply:
Thank you for the response,

In [a test in moonbeam](https://github.com/moonbeam-foundation/moonbeam/blob/c6d07d703dfcdd94cc311fa83b553071b7d433ff/test/suites/dev/moonbase/test-subscription/test-subscription.ts#L20-L31) we found that the `id_provider` is being reset somehow  and changed to the default one. Changing `.take()` to `.clone()` would fix the issue.

I've opened a PR: https://github.com/paritytech/polkadot-sdk/pull/6588 addressing this issue

Reply:
Ok, weird that wasn't obvious to me but your fix looks good to me